src/librustc_middle/ty/mod.rs

```diff
@@ -1571,6 +1570,92 @@ pub type PlaceholderType = Placeholder<BoundVar>;
 
 pub type PlaceholderConst = Placeholder<BoundVar>;
 
+/// A `DefId` which is potentially bundled with its corresponding generic parameter
+/// in case `did` is a const argument.
+///
+/// This is used to prevent cycle errors during typeck
+/// as `type_of(const_arg)` depends on `typeck_tables_of(owning_body)`
+/// which once again requires the type of its generic arguments.
+///
+/// Luckily we only need to deal with const arguments once we
+/// know their corresponding parameters. We (ab)use this by
+/// calling `type_of(param_did)` for these arguments.
+///
+/// ```rust
+/// #![feature(const_generics)]
+///
+/// struct A;
+/// impl A {
+///     fn foo<const N: usize>(&self) -> usize { N }
+/// }
+/// struct B;
+/// impl B {
+///     fn foo<const N: u8>(&self) -> usize { 42 }
+/// }
+///
+/// fn main() {
+///     let a = A;
+///     a.foo::<7>();
+/// }
+/// ```
+#[derive(Copy, Clone, Debug, TypeFoldable, Lift, RustcEncodable, RustcDecodable)]
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Hash, HashStable)]
+pub struct WithOptParam<T> {
+    pub did: T,
```

Comment:
Btw, if you want to use a name for this that isn't `def_id`, I'd prefer `id`. Bonus: when you name the struct `def`, you get `def.id`, which reads better to me than `def.did`.

Reply:
I prefer `did` here. Using `id` makes building a new `WithOptConstParam` less clear imo:
```rust
fn optimized_mir_of_const_arg<'tcx>(
    tcx: TyCtxt<'tcx>,
    (id, param_did): (LocalDefId, DefId),
) -> &'tcx Body<'tcx> {
    tcx.arena.alloc(inner_optimized_mir(
        tcx,
        ty::WithOptConstParam { id, const_param_did: Some(param_did) },
    ))
}
```